core/src/main/scala/no/nrk/bigquery/BQRoutine.scala

```diff
@@ -6,13 +6,78 @@
 
 package no.nrk.bigquery
 
-import cats.data.NonEmptyList
-import cats.syntax.all._
+import cats.syntax.all.*
 import cats.Show
-import no.nrk.bigquery.UDF._
-import no.nrk.bigquery.UDF.UDFId._
-import no.nrk.bigquery.syntax._
-import no.nrk.bigquery.util.{IndexSeqSizedBuilder, Nat, Sized}
+import cats.data.NonEmptyList
+import no.nrk.bigquery.UDF.UDFId
+import no.nrk.bigquery.UDF.UDFId.{PersistentId, TemporaryId}
+import no.nrk.bigquery.syntax.*
+import no.nrk.bigquery.util.IndexSeqSizedBuilder
+import no.nrk.bigquery.util.{Nat, Sized}
+
+sealed trait BQRoutine[N <: Nat] {
+  def params: BQRoutine.Params[N]
+  def call(args: Sized[IndexedSeq[BQSqlFrag.Magnet], N]): BQSqlFrag.Call =
+    BQSqlFrag.Call(this, args.unsized.toList.map(_.frag))
+}
+
+sealed trait BQPersistentRoutine[N <: Nat] extends BQRoutine[N] {
+  def name: BQPersistentRoutine.PersistentRoutineId
+}
+
+object BQPersistentRoutine {
+  trait PersistentRoutineId {
+    def dataset: BQDataset
+    def name: Ident
+    def asString: String
+  }
+}
+
+object BQRoutine {
+  type Params[N <: Nat] = Sized[IndexedSeq[Param], N]
+  object Params extends IndexSeqSizedBuilder[Param]
+
+  case class Param(name: Ident, maybeType: Option[BQType]) {
+    def definition: BQSqlFrag =
+      maybeType match {
+        case Some(tpe) => bqfr"$name $tpe"
+        case None => bqfr"$name ANY TYPE"
+      }
+  }
+  object Param {
+    def apply(name: String, tpe: BQType): Param =
+      Param(Ident(name), Some(tpe))
+
+    def untyped(name: String): Param =
+      Param(Ident(name), None)
+
+    def fromField(field: BQField): Param =
+      Param(Ident(field.name), Some(BQType.fromField(field)))
+  }
+
+}
+
+case class TVF[+P, N <: Nat](
+    name: TVF.TVFId,
+    // this doesn't exist physically, only in a sense when querying
+    partitionType: BQPartitionType[P],
```

Comment:
@eirikhalvard do you remember why we added the `partitionType` here? Looks like it's copied from `View` but it's never used.

Reply:
I think this was an attempt to let TVF be similar to views. One benefits of doing so is that we can use all the other concepts  related to tables and views. However, then we need the result of when we call the TVF to be an instance of `BQTableLike` or maybe `ViewLike`.

Reply:
i don't remember exactly why, but it is probably because of keeping it similar to views as you say